File: algorithms/binary-search/python/binarysearch.py
#!/usr/bin/env python3

def binary_search(arr, val):
  left, right = 0, len(arr) - 1
  while left <= right:
    mid = (left + right) // 2
    if val > arr[mid]:
      left = mid + 1
    elif val < arr[mid]:
      right = mid - 1
    else:
      return True
  return False

def _binary_search_rec(arr, val, lo, hi):
  if lo > hi:
    return False
  mid = (lo + hi) // 2
  if arr[mid] > val:
    return _binary_search_rec(arr, val, lo, mid - 1)
  if arr[mid] < val:
    return _binary_search_rec(arr, val, mid + 1, hi)
  return True

def binary_search_rec(arr, val):
  return _binary_search_rec(arr, val, 0, len(arr) - 1)

File: algorithms/binary-search/c/binarysearch.c
#include <stdio.h>

int binary_search(int *arr, int len, int val)
{
  int lo = 0, hi = len - 1;
  int mid;

  while (lo <= hi) {
    mid = (lo + hi) / 2;
    if (arr[mid] < val) {
      lo = mid + 1;
    } else if (arr[mid] > val) {
      hi = mid - 1;
    } else {
      return 1;
    }
  }
  return 0;
}

int _binary_search_rec(int *arr, int val, int lo, int hi)
{
  if (lo > hi) {
    return 0;
  }

  int mid = (lo + hi) / 2;

  if (arr[mid] < val) {
    return _binary_search_rec(arr, val, mid + 1, hi);
  }

  if (arr[mid] > val) {
    return _binary_search_rec(arr, val, lo, mid - 1);
  }

  return 1;
}

int binary_search_rec(int *arr, int len, int val)
{
  return _binary_search_rec(arr, val, 0, len - 1);
}


